src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs

```diff
@@ -55,6 +55,20 @@ public static bool IsSystemVoid([NotNullWhen(returnValue: true)] this ITypeSymbo
         public static bool IsNullable([NotNullWhen(returnValue: true)] this ITypeSymbol? symbol)
             => symbol?.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T;
 
+        public static bool IsNullable(
+            [NotNullWhen(returnValue: true)] this ITypeSymbol? symbol,
+            [NotNullWhen(returnValue: true)] out ITypeSymbol? underlyingType)
+        {
+            if (symbol is INamedTypeSymbol { OriginalDefinition: { SpecialType: SpecialType.System_Nullable_T } } namedType)
```

Comment:
fwiw: my comments are often nits.  it's more of a general: here's how i feel.  not a "you must change this" :)  I'm finding with patterns that something doesn't quite feel great just yet, so i'm hoping these discussions can help us land in a good place.

Note: i think my ideal preference (not as part of this PR) would be that this function was more like:

```c#
public static (bool result, ITypeSymbol underlyingType) IsNullable(this ITypeSymbol type)
    => type switch
    {
        INamedTypeSymbol { OriginalDefinition: { SpecialType: SpecialType.System_Nullable_T }, TypeArguments: var args } => (true, args[0])
        _ => default;
    };
```

Or:

```c#
public static (bool result, ITypeSymbol underlyingType) IsNullable(this ITypeSymbol type)
    => type switch
    {
        INamedTypeSymbol
        {
            OriginalDefinition: { SpecialType: SpecialType.System_Nullable_T },
            TypeArguments: var args,
        } => (true, args[0])
        _ => default;
    };
```

Basically, i don't mind going heavily into patterns in expression-oriented code.  It's when we have the mix of statements and patterns that i find things feel very clunky confusing.

Reply:
>  it's more of a general: here's how i feel. not a "you must change this"

Right, and it's up to me to figure out how to weight your feedback. I usually assume you're bringing in project-specific context that I don't have. :D

> Basically, i don't mind going heavily into patterns in expression-oriented code. It's when we have the mix of statements and patterns that i find things feel very clunky confusing.

I don't see anything clunky per se in mixing paradigms like that. Is it at the conceptual level that it feels clunky to you or at the syntactic level?

Reply:
> Is it at the conceptual level that it feels clunky to you or at the syntactic level?

A bit of both.  Conceptually, mixing patterns and statements in this manner doesn't feel "good" for me.  They feel a little too far apart to do this.  It doesn't help the syntax then feels pretty clunky here.  

Expressions + patterns seems to feel like it just meshes better.

Expressions often feel lightweight.  So mixing in either a lightweight or heavyweight pattern doesn't feel too bad.  Statements feel heavier at times, so mixing in a lightweight expression or lightweight pattern is fine.   HOwever, mixing heavy statements and heavy expressions/patterns just feels like too much.

Note: this is why i will commonly push back on `if-statements` with extremely heavy condition expressions (i.e. not even containing patterns).

Does that make any sort of sense?

Reply:
It's because it makes you go too far into a separate mode of reasoning when one or the other is "heavy"? Yes, that makes sense. The difference between us is that my bar for what constitutes "heavy" is higher than yours. Do you think there's a temporary element of unfamiliarity adding to the effective heaviness?

Reply:
> The difference between us is that my bar for what constitutes "heavy" is higher than yours.

Makes sense

>  Do you think there's a temporary element of unfamiliarity adding to the effective heaviness?

Yes.  Though i also think it's temporally due to the ways these syntactic structures have been designed over time.  Our more 'modern' constructs feel like tehy work better together to me for these types of scenarios.  Mixing matching some of the oldest and some of the newest (when both are bulky) feels the most 'off'.  

> It's because it makes you go too far into a separate mode of reasoning when one or the other is "heavy"?

Somewhat, yes. With a pattern i'm strongly thinking about the matching of data to the pattern (and how it deconstructs). `if` statements are then much more about flow control and are generally about lightly considering a condition which can then flow into heavy false/branches.  So it's a mixing of something dense (the pattern logic) into an area i would generally prefer to be light.

Reply:
Also, just syntactically, this stuff doesn't feel like it meshes well.  And i think that clunkyness becomes distracting for me and pulls me out from being able to concentrate and understand the code well.